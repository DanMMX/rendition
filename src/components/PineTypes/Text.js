const RegexParser = require('regex-parser')
import * as React from 'react'
import Textarea from '../Textarea'
import * as showdown from 'showdown'
const isString = require('lodash/isString')

const converter = new showdown.Converter()

const normalizeString = string => {
  return string || ''
}

export const rules = {
  is: (target, value) => normalizeString(target) === value,
  contains: (target, value) => normalizeString(target).includes(value),
  'does not contain': (target, value) =>
    !normalizeString(target).includes(value),
  'matches RegEx': (target, value) =>
    normalizeString(target).match(RegexParser(value)),
  'does not match RegEx': (target, value) =>
    !normalizeString(target).match(RegexParser(value))
}

export const validate = isString

export const normalize = value => (value == null ? '' : value.toString())

export const Edit = ({ onChange, ...props }) => (
  <Textarea onChange={e => onChange(e.target.value)} {...props} />
)

export const Display = ({ data, ...props }) => (
  <div
    {...props}
    className='markdown-body'
    dangerouslySetInnerHTML={{ __html: converter.makeHtml(data) }}
  />
)
