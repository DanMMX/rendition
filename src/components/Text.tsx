import { TextProps, Theme } from 'rendition';
import styled from 'styled-components';
import { color, fontSize, responsiveStyle, space } from 'styled-system';
import asRendition from '../asRendition';
import { monospace } from '../utils';

interface ThemedTextProps extends TextProps {
	theme: Theme;
}

export const caps = (props: ThemedTextProps) =>
	props.caps
		? {
				textTransform: 'uppercase',
				letterSpacing: '0.2em',
			}
		: null;

export const bold = (props: ThemedTextProps) =>
	props.bold
		? { fontWeight: props.theme.weights[props.theme.weights.length - 1] }
		: null;

export const align = responsiveStyle('text-align', 'align');

const Text = styled.div`
	${align}
	${color}
	${fontSize}
	${monospace as any};
	${space}

	${caps as any}
	${bold as any}
`;

const Base = asRendition(Text) as React.ComponentClass<TextProps> & {
	span: React.ComponentClass<TextProps>;
	p: React.ComponentClass<TextProps>;
};
Base.displayName = 'Text';
Base.span = asRendition(Text.withComponent('span'));
Base.p = asRendition(Text.withComponent('p'));

export default Base;
