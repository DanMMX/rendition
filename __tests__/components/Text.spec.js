import React from 'react';
import renderer from 'react-test-renderer';
import Provider from '../../src/components/Provider'
import Text from '../../src/components/Text'

test('Text renders correctly', () => {
  const component = renderer.create(
    <Provider>
      <Text>Text</Text>
    </Provider>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
