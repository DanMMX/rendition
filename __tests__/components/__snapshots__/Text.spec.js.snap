// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text renders correctly 1`] = `
<div
  className="sc-bdVaJa jKopLr"
>
  <div
    className="sc-htoDjs gDCMNh sc-gzVnrw faVbst"
  >
    Text
  </div>
</div>
`;
